### PR TITLE
rewrite spoke cluster creating controller

### PR DESCRIPTION
**Pull request by skeeey**

### CODE REVIEW

**Comment by pmorie** [comment]
looks good.

/approve

File: pkg/spoke/spokeagent.go [hunk 1]
```diff
@@ -90,6 +92,39 @@ func (o *SpokeAgentOptions) RunSpokeAgent(ctx context.Context, controllerContext
...
 
+	// load bootstrap client config
+	bootstrapClientConfig, err := clientcmd.BuildConfigFromFlags("", o.BootstrapKubeconfig)
+	if err != nil {
+		return fmt.Errorf("unable to load bootstrap kubeconfig from file %q: %w", o.BootstrapKubeconfig, err)
+	}
+
+	// start a SpokeClusterCreatingController to make sure there is a spoke cluster on hub cluster
+	// if the bootstrap client config is valid
+	if isBootstrapClientConfigValid(bootstrapClientConfig) {
```
Comment (deads2k) [inline pkg/spoke/spokeagent.go]: let's just require it by not testing this.  We can require well-formedness

File: pkg/spoke/spokeagent.go [hunk 2]
```diff
@@ -90,6 +92,39 @@ func (o *SpokeAgentOptions) RunSpokeAgent(ctx context.Context, controllerContext
...
+		bootstrapClusterClient, err := clusterv1client.NewForConfig(bootstrapClientConfig)
+		if err != nil {
+			return err
+		}
+
+		caBundle := controllerContext.KubeConfig.CAData
```
Comment (deads2k) [inline pkg/spoke/spokeagent.go]: factor into a method please.

File: pkg/spoke/spokecluster/creating_controller.go [hunk 3]
```diff
@@ -28,50 +27,50 @@ func NewSpokeClusterCreatingController(
...
-	spokeCluster := &clusterv1.SpokeCluster{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: c.clusterName,
-		},
-		Spec: clusterv1.SpokeClusterSpec{
-			SpokeClientConfig: clusterv1.ClientConfig{
-				URL:      c.spokeExternalServerUrl,
-				CABundle: c.spokeCABundle,
+	_, err := c.hubClusterClient.ClusterV1().SpokeClusters().Get(ctx, c.clusterName, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
```
Comment (deads2k) [inline pkg/spoke/spokecluster/creating_controller.go]:
```
switch
case unauthorized
    v4
case forbidden && error contains anonymous
    v4
case isNotFound
    fall to below
case err == nil
    return
case err != nil
    return err
```

Then this block isn't nested

File: pkg/spoke/spokeagent.go (hunk 2 above)
Comment (skeeey) [reply pkg/spoke/spokeagent.go]: done

File: pkg/spoke/spokecluster/creating_controller.go (hunk 3 above)
Comment (skeeey) [reply pkg/spoke/spokecluster/creating_controller.go]: done

File: pkg/spoke/spokeagent.go (hunk 1 above)
Comment (skeeey) [reply pkg/spoke/spokeagent.go]: remove the testing code

File: pkg/spoke/spokecluster/creating_controller.go [hunk 4]
```diff
@@ -28,24 +31,31 @@ func NewSpokeClusterCreatingController(
...
 		clusterName:            clusterName,
 		spokeExternalServerUrl: spokeExternalServerUrl,
 		spokeCABundle:          spokeCABundle,
-		hasSpokeClusterFunc:    hasSpokeClusterFunc,
 		hubClusterClient:       hubClusterClient,
 	}
 	return factory.New().
 		WithSync(c.sync).
-		ResyncEvery(5*time.Minute).
+		ResyncEvery(60*time.Minute).
```
Comment (deads2k) [inline pkg/spoke/spokecluster/creating_controller.go:43]: we may have to tighten this up later.  Don't be surprised at that outcome, but its good to start.

**Pull request merged**